src/sqlfluff/core/templaters/python.py

```diff
@@ -249,9 +250,23 @@ def process(
         live_context = self.get_context(fname=fname, config=config)
 
         def render_func(raw_str: str) -> str:
-            """Render the string using the captured live_context."""
+            """Render the string using the captured live_context.
+
+            In order to support mocking of template variables
+            containing "." characters, this function converts any
+            template variable containing "." into a dictionary lookup.
+                Example:  {foo.bar} => {sqlfluff[foo.bar]}
+            """
             try:
-                rendered_str = raw_str.format(**live_context)
+                # Hack to allow template variables with dot notation (e.g. foo.bar)
+                raw_str_with_dot_notation_hack = re.sub(
+                    r"{([^:}]*\.[^:}]*)(:\S*)?}", r"{sqlfluff[\1]\2}", raw_str
+                )
+                templater_logger.debug(
+                    "    Raw String with Dot Notation Hack: %r",
+                    raw_str_with_dot_notation_hack,
+                )
+                rendered_str = raw_str_with_dot_notation_hack.format(**live_context)
             except KeyError as err:
                 raise SQLTemplaterError(
```

Comment:
@timchurch - can you check how your solution _fails_ if the user _has_ a variable containing a dot (e.g. `{foo.bar}`), but _hasn't defined the sqlfluff dict object at all_?

I think we need to account for that failure mechanism here and provide a hint to the user on how to resolve it. Especially given this will be a new feature, there's a chance that people might not know it's an option unless it's in an error message.

We might want to consider adding an _empty_ dictionary to the default config if that changes the behaviour, so that there's _always_ a `sqlfluff` value present even if the user hasn't set one. That also provides an opportunity to link to the docs in the default config. I'm less confident on this last point though, I think it depends more how the code fails.

Reply:
This is the current error message if sqlfluff dict object is missing from context:

```python
sqlfluff.core.errors.SQLTemplaterError: Failure in Python templating: 'sqlfluff'. Have you configured your variables? https://docs.sqlfluff.com/en/stable/configuration.html#templating-configuration
```

But I agree that this is confusing for users and worth improving.  I'll add a custom error message here.

Reply:
I've updated the error messages but still open to suggestions.

Here is the initial error message if `sqlfluff` key does not exist:
```
sqlfluff.core.errors.SQLTemplaterError: Failure in Python templating: magic key 'sqlfluff' missing from context.  This key is required for template variables containing '.'. https://docs.sqlfluff.com/en/stable/perma/python_templating.html
```

And here is the error message if `sqlfluff` key does exist but is missing an individual key:
```
sqlfluff.core.errors.SQLTemplaterError: Failure in Python templating: 'foo.bar' key missing from 'sqlfluff' dict in context. Template variables containing '.' are required to use the 'sqlfluff' magic fixed context key. https://docs.sqlfluff.com/en/stable/perma/python_templating.html
```

Reply:
> We might want to consider adding an empty dictionary to the default config if that changes the behaviour, so that there's always a sqlfluff value present even if the user hasn't set one. That also provides an opportunity to link to the docs in the default config. I'm less confident on this last point though, I think it depends more how the code fails.

I'm also open to this.  I briefly tested and seems easy to implement.  However, I'm not sure if it actually helps very much because end user will still need to define this key in their own context anyway for templating to work.